pkg/storage/stores/shipper/uploads/table_manager.go

```diff
@@ -0,0 +1,230 @@
+package uploads
+
+import (
+	"context"
+	"errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"sync"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	"github.com/cortexproject/cortex/pkg/chunk/local"
+	chunk_util "github.com/cortexproject/cortex/pkg/chunk/util"
+	pkg_util "github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type Config struct {
+	Uploader       string
+	IndexDir       string
+	UploadInterval time.Duration
+}
+
+type TableManager struct {
+	cfg             Config
+	boltIndexClient BoltDBIndexClient
+	storageClient   StorageClient
+
+	metrics   *metrics
+	tables    map[string]*Table
+	tablesMtx sync.RWMutex
+
+	done chan struct{}
+	wg   sync.WaitGroup
+}
+
+func NewTableManager(cfg Config, boltIndexClient BoltDBIndexClient, storageClient StorageClient, registerer prometheus.Registerer) (*TableManager, error) {
+	tm := TableManager{
+		cfg:             cfg,
+		boltIndexClient: boltIndexClient,
+		storageClient:   storageClient,
+		metrics:         newMetrics(registerer),
+		done:            make(chan struct{}),
+	}
+
+	tables, err := tm.loadTables()
+	if err != nil {
+		return nil, err
+	}
+
+	tm.tables = tables
+	return &tm, nil
+}
+
+func (tm *TableManager) loop() {
+	defer tm.wg.Done()
+
+	syncTicker := time.NewTicker(tm.cfg.UploadInterval)
+	defer syncTicker.Stop()
+
+	for {
+		select {
+		case <-syncTicker.C:
+			err := tm.uploadTables(context.Background())
```

Comment:
I don't think you should be using a context.Background here.

Reply:
What is the issue when you cancel this ?

Reply:
I don't think we would ever want to cancel that context, not even while stopping to make sure we stop the service only after uploading the files. We could have a context with a minimum of 10 mins timeout to avoid waiting on it forever. What do you think? Is that what you are looking for?

Reply:
My problem is mostly when using context.Background you're blocking the waitgroup and you'll likely be killed and you don't know when. I'm wondering if this could be a problem ? Basically being stopped before gracefully stopping.

First I think you should cancel the loop because the upload happen again anyways after stop. Am I wrong ?

Second 10min would get just shutdown by Kubernetes see https://cloud.google.com/blog/products/gcp/kubernetes-best-practices-terminating-with-grace

> At this point, Kubernetes waits for a specified time called the termination grace period. By default, this is 30 seconds. It’s important to note that this happens in parallel to the preStop hook and the SIGTERM signal. Kubernetes does not wait for the preStop hook to finish.

Basically you're going to get SIGKILL after 30s or more depending on what we have configured, for sure not 10min.

Let's change the Close function to give a 1m timeout, and log if it has been cancelled or success, so we can see if we get killed before finishing.

Reply:
If I am not wrong our ingesters don't get killed within 30s of receiving SIGTERM because we do chunk transfers and it could take a couple of minutes. I checked the config and it seems we configure termination grace period to 80 minutes, see https://github.com/grafana/loki/blob/a0dc10a03a0a82fef7e4783ceeac0fecd65a0c58/production/ksonnet/loki/ingester.libsonnet#L31

Reply:
I think this context should have a timeout and should be configured to a value less than the termination grace period and we should have explicit info level logging on when a table upload starts and completed.

I don't like the idea of hiding a table upload into the background without any visibility, many people run Loki with much more aggressive shutdown requirements and it shoudl be clear from the logs if table uploads succeeded or if the process was killed before they completed, have a timeout less than the grace period would ensure we log that upload filed before we gave up and shutdown.

Reply:
After some more discussion with @sandeepsukhani , we already try to flush chunks forever on shutdown it makes sense to try to upload tables forever on shutdown.

I would like to see the explicit logging of both started and succeeded (as well as error) so that someone can determine from their logs if all tables were uploaded.